SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift

```diff
@@ -0,0 +1,214 @@
+///    %x = alloc_ref [stack] $X
+///      ...
+///    set_deallocating %x
+///    %d = function_ref @dealloc_of_X
+///    %a = apply %d(%x)
+///    dealloc_stack_ref %x
+///
+/// The optimization is only done for stack promoted objects because they are
+/// known to have no associated objects (which are not explicitly released
+/// in the deinit method).
+let releaseDevirtualizerPass = FunctionPass(
+  name: "release-devirtualizer", { function, context in
+    for block in function.blocks {
+      // The last `release_value`` or `strong_release`` instruction before the
+      // deallocation.
+      var lastRelease: RefCountingInst?
+
+      for instruction in block.instructions {
+        if let release = lastRelease {
+          // We only do the optimization for stack promoted object, because for
+          // these we know that they don't have associated objects, which are
+          // _not_ released by the deinit method.
+          if let deallocStackRef = instruction as? DeallocStackRefInst {
+            tryDevirtualizeReleaseOfObject(context, release, deallocStackRef)
+            lastRelease = nil
+            continue
+          }
+        }
+
+        if instruction is ReleaseValueInst || instruction is StrongReleaseInst {
+          lastRelease = instruction as? RefCountingInst
+        } else if instruction.mayRelease || instruction.mayReadRefCount {
+          lastRelease = nil
+        }
+      }
+    }
+  }
+)
+
+/// Tries to de-virtualize the final release of a stack promoted object.
+private func tryDevirtualizeReleaseOfObject(
+  _ context: PassContext,
+  _ release: RefCountingInst,
+  _ deallocStackRef: DeallocStackRefInst
+) {
+  let allocRefInstruction = deallocStackRef.allocRef
+  var root = release.operands[0].value
+  while let newRoot = stripRCIdentityPreservingInsts(root) {
+    root = newRoot
+  }
+
+  if root != allocRefInstruction {
+    return
+  }
+
+  let type = allocRefInstruction.type
+
+  guard let dealloc = context.getDestructor(ofClass: type) else {
+    return
+  }
+
+  let builder = Builder(at: release, location: release.location, context)
+
+  var object: Value = allocRefInstruction
+  if object.type != type {
+    object = builder.createUncheckedRefCast(object: object, type: type)
+  }
+
+  // Do what a release would do before calling the deallocator: set the object
+  // in deallocating state, which means set the RC_DEALLOCATING_FLAG flag.
+  builder.createSetDeallocating(operand: object, isAtomic: release.isAtomic)
+
+  // Create the call to the destructor with the allocated object as self
+  // argument.
+  let functionRef = builder.createFunctionRef(dealloc)
+
+  let substitutionMap = context.getContextSubstitutionMap(for: type)
+  builder.createApply(function: functionRef, substitutionMap, arguments: [object])
+  context.erase(instruction: release)
+}
+
+private func stripRCIdentityPreservingInsts(_ value: Value) -> Value? {
+  switch value {
+  // First strip off RC identity preserving casts.
```

Comment:
If you do a `guard let inst = value as? Instruction else { return }` before the switch, you can switch over `inst` which simplifies the code below:
``` 
case is UpcastInst, is UncheckedRefCastInst, ...
```